File: mersenne_twister.py
import pandas as pd
import random
import sys


def getDecimalFractions(n_samples: int) -> list:
    return [random.random() for _ in range(n_samples)]


if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: python mersenne_twister.py {n_samples}")
        sys.exit(1)

    n_samples = int(sys.argv[1])
    mersenne_twister = getDecimalFractions(n_samples)

    df = pd.DataFrame(mersenne_twister, columns=["Mersenne-Twister Numbers"])
    df.to_csv("data/mersenne_twister.csv", index=False)

File: mersenne_twister.cpp
#include <bits/stdc++.h>
#include <ctime>
#include <random>

std::vector<long double> getDecimalFractions(unsigned long n_samples) {
    std::mt19937_64 gen(time(nullptr));
    std::vector<long double> result(n_samples);

    for (unsigned long i = 0; i < n_samples; i++) {
        result[i] = gen() / (long double)gen.max();
    }

    return result;
}

int main(int argc, char *argv[]) {
    if (argc < 2) {
        std::cout << "Usage: ./mersenne_twister {n_samples}" << std::endl;
        return 1;
    }

    unsigned long n_samples = std::strtoul(argv[1], nullptr, 10);
    std::vector<long double> result = getDecimalFractions(n_samples);

    std::cout << std::fixed << std::setprecision(20);
    std::cout << "Mersenne-Twister Numbers" << std::endl;
    for (const auto &num : result) {
        std::cout << num << std::endl;
    }

    return 0;
}
